from pathlib import Path
from PyQt5 import QtWidgets, QtGui, QtCore, uic
import cv2

from package.image_utils import to_pixmap, draw_points, highlight_rgb, make_pixel_map
from package.color_utils import add_color_def, save_defs, clear_defs
from package.operation import (
    DRAW_POINT_RADIUS, DRAW_POINT_LIMIT, UI_UPDATE_INTERVAL,
    SPHERE_RADIUS, PICTURE_DIR
)

UI_FILE = Path(__file__).resolve().with_name("mainwindow.ui")


class PhotoViewer(QtWidgets.QDialog):
    def __init__(self, parent=None):
        super().__init__(parent)
        uic.loadUi(str(UI_FILE), self)

        # === ì™¼ìª½(real_photo) : ì›ë³¸ ===
        self.scene = QtWidgets.QGraphicsScene(self)
        self.real_photo.setScene(self.scene)
        self.pixmap_item = None
        self.current_img = None

        # === ì˜¤ë¥¸ìª½(pixel_view) : ë¶„ë¥˜ ê²°ê³¼ ===
        self.pixel_scene = QtWidgets.QGraphicsScene(self)
        self.pixel_view.setScene(self.pixel_scene)
        self.pixelmap_item = None
        self.current_pixel_map = None  # í˜„ì¬ í”½ì…€ë§µ ì €ì¥ (ìš°ì¸¡ ë™ê¸°í™”ìš©)

        self.files = self._scan_files()
        self.index = 0

        # ë²„íŠ¼ ì—°ê²°
        self.clearButton.clicked.connect(self.clear_folder)
        self.nextButton.clicked.connect(self.next_photo)
        self.saveButton.clicked.connect(self.confirm_colors)
        self.exitButton.clicked.connect(self.safe_exit)
        self.clearDataButton.clicked.connect(self.clear_data)

        # ì£¼ê¸°ì  ê°±ì‹ (ì‹ ê·œ íŒŒì¼ ê°ì§€)
        self.timer = QtCore.QTimer(self)
        self.timer.timeout.connect(self.update_photos)
        self.timer.start(UI_UPDATE_INTERVAL)

        if self.files:
            self.show_photo(self.files[self.index])
        else:
            self._show_message("í´ë”ê°€ ë¹„ì–´ ìˆìŠµë‹ˆë‹¤")

        # ë“œë¡œì‰ ê´€ë ¨(ì™¼ìª½ì—ì„œë§Œ ë“œë˜ê·¸)
        self.drawing = False
        self.selected_points = []
        self.pending_colors = {}  # ì„ì‹œ RGB ì €ì¥
        self.real_photo.viewport().installEventFilter(self)

        # main.pyì—ì„œ ì£¼ì… ê°€ëŠ¥
        self.cap_proc = None

    # -------------------------------
    def _scan_files(self):
        PICTURE_DIR.mkdir(parents=True, exist_ok=True)
        return sorted(PICTURE_DIR.glob("frame_*.jpg"))

    def _show_message(self, text: str):
        self.scene.clear()
        self.scene.addText(text, QtGui.QFont("Arial", 14))

    # === ì˜¤ë¥¸ìª½ ë·° ê°±ì‹  í—¬í¼ ===
    def update_pixel_view(self):
        if self.current_img is None:
            self.pixel_scene.clear()
            return
        pixel_map = make_pixel_map(self.current_img)
        pixmap2 = to_pixmap(pixel_map, QtGui)
        self.pixel_scene.clear()
        self.pixelmap_item = self.pixel_scene.addPixmap(pixmap2)
        self.pixel_view.fitInView(self.pixelmap_item, QtCore.Qt.KeepAspectRatio)

    def show_photo(self, fpath: Path):
        img = cv2.imread(str(fpath))
        if img is None:
            self._show_message(f"ì´ë¯¸ì§€ë¥¼ ë¶ˆëŸ¬ì˜¬ ìˆ˜ ì—†ìŠµë‹ˆë‹¤:\n{fpath.name}")
            return
        self.current_img = img

        # ì™¼ìª½: ì›ë³¸
        pixmap = to_pixmap(img, QtGui)
        self.scene.clear()
        self.pixmap_item = self.scene.addPixmap(pixmap)
        self.pixmap_item.setAcceptedMouseButtons(QtCore.Qt.NoButton)
        self.real_photo.fitInView(self.pixmap_item, QtCore.Qt.KeepAspectRatio)

        # ì˜¤ë¥¸ìª½: ë¶„ë¥˜ ê²°ê³¼
        self.update_pixel_view()

    def next_photo(self):
        self.files = self._scan_files()
        if not self.files:
            self._show_message("í´ë”ê°€ ë¹„ì–´ ìˆìŠµë‹ˆë‹¤")
            return
        self.index = (self.index + 1) % len(self.files)
        self.show_photo(self.files[self.index])

    def clear_folder(self):
        for f in PICTURE_DIR.glob("frame_*.jpg"):
            try:
                f.unlink()
            except Exception:
                pass
        self.files, self.index = [], 0
        self._show_message("í´ë”ê°€ ë¹„ì–´ ìˆìŠµë‹ˆë‹¤")
        # ì˜¤ë¥¸ìª½ë„ ì´ˆê¸°í™”
        self.pixel_scene.clear()

    def update_photos(self):
        new_files = self._scan_files()
        if new_files != self.files:
            self.files = new_files

    def resizeEvent(self, event):
        super().resizeEvent(event)
        if self.files and self.pixmap_item:
            self.real_photo.fitInView(self.pixmap_item, QtCore.Qt.KeepAspectRatio)
        if self.pixelmap_item:
            self.pixel_view.fitInView(self.pixelmap_item, QtCore.Qt.KeepAspectRatio)

    # -------------------------------
    def get_selected_label(self):
        if self.product.isChecked():
            return "product"
        elif self.defect.isChecked():
            return "defect"
        elif self.background.isChecked():
            return "background"
        return None

    def confirm_colors(self):
        """Save ë²„íŠ¼ â†’ ì„ì‹œ RGBë¥¼ Sphereë¡œ ë“±ë¡í•˜ê³  ì €ì¥ + ì˜¤ë¥¸ìª½ ì¦‰ì‹œ ê°±ì‹ """
        for label, rgb_set in self.pending_colors.items():
            if rgb_set:
                for rgb in rgb_set:
                    add_color_def(label, rgb, radius=SPHERE_RADIUS)
                print(f"[{label}] {len(rgb_set)}ê°œ RGB â†’ Sphereë¡œ ë“±ë¡ë¨")
        self.pending_colors.clear()

        save_defs()
        print("color_defs.jsonì— ì €ì¥ ì™„ë£Œ âœ…")

        # ì˜¤ë¥¸ìª½ ë·° ì¦‰ì‹œ ê°±ì‹ 
        self.update_pixel_view()

    def clear_data(self):
        """Data Clear â†’ JSON ì´ˆê¸°í™” + ì˜¤ë¥¸ìª½ ì¦‰ì‹œ ê°±ì‹ """
        clear_defs()
        QtWidgets.QMessageBox.information(self, "Data Clear", "ì €ì¥ëœ ìƒ‰ìƒ ì •ì˜ê°€ ëª¨ë‘ ì‚­ì œë˜ì—ˆìŠµë‹ˆë‹¤ âœ…")
        self.update_pixel_view()

    def safe_exit(self):
        print("ğŸ”’ ì•ˆì „ ì¢…ë£Œ ì‹œì‘")
        save_defs()
        print("âœ… ìƒ‰ìƒ ì •ì˜ ì €ì¥ ì™„ë£Œ")

        if self.cap_proc and self.cap_proc.poll() is None:
            self.cap_proc.terminate()
            print("ğŸ“· ìº¡ì³ í”„ë¡œì„¸ìŠ¤ ì¢…ë£Œ")

        QtWidgets.QApplication.quit()

    # -------------------------------
    def eventFilter(self, source, event):
        if source == self.real_photo.viewport():
            if event.type() == QtCore.QEvent.MouseButtonPress:
                if event.button() == QtCore.Qt.LeftButton:
                    self.drawing = True
                    self.selected_points = []
                    return True

            elif event.type() == QtCore.QEvent.MouseMove:
                if self.drawing and self.current_img is not None:
                    pos = self.real_photo.mapToScene(event.pos()).toPoint()
                    x, y = pos.x(), pos.y()
                    h, w = self.current_img.shape[:2]
                    if 0 <= x < w and 0 <= y < h:
                        self.selected_points.append((x, y))
                        overlay = draw_points(
                            self.current_img,
                            self.selected_points[-DRAW_POINT_LIMIT:],
                            radius=DRAW_POINT_RADIUS
                        )
                        pixmap = to_pixmap(overlay, QtGui)
                        self.scene.clear()
                        self.pixmap_item = self.scene.addPixmap(pixmap)
                    return True

            elif event.type() == QtCore.QEvent.MouseButtonRelease:
                if event.button() == QtCore.Qt.LeftButton and self.current_img is not None:
                    self.drawing = False
                    label = self.get_selected_label()
                    if not label:
                        print("ë¼ë²¨ì´ ì„ íƒë˜ì§€ ì•Šì•˜ìŠµë‹ˆë‹¤.")
                        return True

                    # ë“œë˜ê·¸ êµ¬ê°„ RGB ìˆ˜ì§‘
                    rgb_set = set()
                    img_rgb = cv2.cvtColor(self.current_img, cv2.COLOR_BGR2RGB)
                    for (x, y) in self.selected_points:
                        rgb_set.add(tuple(img_rgb[int(y), int(x)]))

                    if label not in self.pending_colors:
                        self.pending_colors[label] = set()
                    self.pending_colors[label].update(rgb_set)

                    print(f"[{label}] {len(rgb_set)}ê°œ RGB ì„ì‹œ ì €ì¥ë¨")

                    # ì™¼ìª½ í•˜ì´ë¼ì´íŠ¸
                    overlay = highlight_rgb(self.current_img, rgb_set)
                    pixmap = to_pixmap(overlay, QtGui)
                    self.scene.clear()
                    self.pixmap_item = self.scene.addPixmap(pixmap)
                    return True
        return False
